subsys/shell/shell_utils.c

```diff
@@ -498,15 +498,16 @@ void z_shell_cmd_trim(const struct shell *sh)
 
 static const struct device *shell_device_internal(size_t idx,
 						  const char *prefix,
-						  shell_device_filter_t filter)
+						  shell_device_filter_t filter,
+						  bool ready)
 {
 	size_t match_idx = 0;
 	const struct device *dev;
 	size_t len = z_device_get_all_static(&dev);
 	const struct device *dev_end = dev + len;
 
 	while (dev < dev_end) {
-		if (device_is_ready(dev)
+		if (((ready && device_is_ready(dev)) || !ready)
```

Comment:
If you did flip the logic of `ready` to be a `show_all` flag:
```suggestion
		if ((show_all || device_is_ready(dev))
```